src/tunnel.c

```diff
@@ -1149,7 +1154,88 @@ int ssl_connect(struct tunnel *tunnel)
 			goto err_ssl_context;
 		}
 	} else { /* end PKCS11 engine */
-#endif
+#else // OPENSSL_VERSION_NUMBER >= 0x30000000L
+    /* OpenSSL 3.0+ provider-based PKCS#11 support */
+    if (tunnel->config->use_engine > 0) {
+        // Debug: Print the certificate path/URI being used
+        log_debug_details("Attempting to load certificate from: %s\n", tunnel->config->user_cert);
+        
+        // Use OSSL_STORE to load certificate and private key from PKCS#11
+        OSSL_STORE_CTX *store_ctx = OSSL_STORE_open(tunnel->config->user_cert, NULL, NULL, NULL, NULL);
+        if (!store_ctx) {
+            log_error("PKCS11 OSSL_STORE_open failed: %s\n",
+                      ERR_error_string(ERR_peek_last_error(), NULL));
+            goto err_ssl_context;
+        }
+
+        X509 *cert = NULL;
+        EVP_PKEY *pkey = NULL;
+        OSSL_STORE_INFO *info;
+
+        // Load all objects from the store
+        while ((info = OSSL_STORE_load(store_ctx)) != NULL) {
+            int type = OSSL_STORE_INFO_get_type(info);
+            
+            if (type == OSSL_STORE_INFO_CERT && !cert) {
```

Comment:
Sorry for the noob question, but can it happen that there are multiple keys in the store? If so, why chose the first one?